File: src/js/3_StructuralPatterns/7_Proxy/proxy_exercise.js
class Person
{
  constructor(age=0)
  {
    this.age = age;
  }

  drink() { return 'drinking'; }
  drive() { return 'driving'; }
  drinkAndDrive() { return 'driving while drunk'; }
}

class ResponsiblePerson
{
  constructor(person)
  {
    this.person = person;
  }

  get age()
  {
    return this.person.age;
  }

  set age(value)
  {
    this.person.age = value;
  }

  drink()
  {
    if (this.age >= 18)
      return this.person.drink();
    return 'too young';
  }

  drive()
  {
    if (this.age >= 16)
      return this.person.drive();
    return 'too young';
  }

  // don't do it, folks!
  drinkAndDrive()
  {
    return 'dead';
  }
}

describe('proxy', function()
{
  it('does its proxy thing', function()
  {
    let p = new Person(10);
    let rp = new ResponsiblePerson(p);

    expect(rp.drive()).toEqual('too young');
    expect(rp.drink()).toEqual('too young');
    expect(rp.drinkAndDrive()).toEqual('dead');

    rp.age = 20;

    expect(rp.drive()).toEqual('driving');
    expect(rp.drink()).toEqual('drinking');
    expect(rp.drinkAndDrive()).toEqual('dead');
  });
});

File: src/java/3_StructuralPatterns/6_Proxy/ProxyExercise.java
package com.activemesa.structural.proxy.exercise;

class Person
{
  private int age;

  public Person(int age)
  {
    this.age = age;
  }

  public int getAge()
  {
    return age;
  }

  public void setAge(int age)
  {
    this.age = age;
  }

  public String drink() { return "drinking"; }
  public String drive() { return "driving"; }
  public String drinkAndDrive() { return "driving while drunk"; }
}

class ResponsiblePerson
{
  private Person person;

  public ResponsiblePerson(Person person)
  {
    this.person = person;
  }

  public void setAge(int value) { person.setAge(value); }
  public int getAge() { return person.getAge(); }

  public String drink()
  {
    return getAge() >= 18 ? person.drink() : "too young";
  }

  public String drive()
  {
    return getAge() >= 16 ? person.drive() : "too young";
  }

  public String drinkAndDrive()
  {
    return "dead";
  }
}